include/dxc/DxilContainer/DxilPipelineStateValidation.h

```diff
@@ -403,6 +424,116 @@ class DxilPipelineStateValidation
   {
   }
 
+  enum class RWMode {
+    Read,
+    CalcSize,
+    Write,
+  };
+
+  class CheckedReaderWriter {
+  private:
+    char *Ptr;
+    uint32_t Size;
+    uint32_t Offset;
+    RWMode Mode;
+
+  public:
+    CheckedReaderWriter(const void *ptr, uint32_t size, RWMode mode)
+      : Ptr(reinterpret_cast<char*>(const_cast<void*>(ptr))),
+        Size(mode == RWMode::CalcSize ? 0 : size), Offset(0), Mode(mode) {}
+
+    uint32_t GetSize() { return Size; }
+    RWMode GetMode() { return Mode; }
+
+    // Return true if size fits in remaing buffer.
+    bool CheckBounds(size_t size) {
+      if (Mode != RWMode::CalcSize) {
+        PSV_RETB(size <= UINT_MAX);
+        PSV_RETB(Offset <= Size);
+        return (uint32_t)size <= Size - Offset;
+      }
+      return true;
+    }
+    // Increment Size by size. Return false on error.
+    bool IncrementPos(size_t size) {
+      PSV_RETB(size <= UINT_MAX);
+      uint32_t uSize = size;
+      if (Mode == RWMode::CalcSize) {
+        PSV_RETB(uSize <= Size + uSize);
+        Size += uSize;
+      }
+      CheckBounds(size);
+      Offset += size;
+      return true;
+    }
+
+    // Cast and return true if it fits.
+    template <typename _T>
+    bool Cast(_T **ppPtr, size_t size) {
+      PSV_RETB(CheckBounds(size));
+      *ppPtr = reinterpret_cast<_T*>(Ptr + Offset);
+      return true;
+    }
+    template <typename _T>
+    bool Cast(_T **ppPtr) {
+      return Cast(ppPtr, sizeof(_T));
+    }
+
+    // Map* methods increment Offset.
+
+    // Assign pointer, increment Offset, and return true if size fits.
+    template<typename _T>
+    bool MapPtr(_T **ppPtr, size_t size = 0) {
+      Cast(ppPtr, size);
+      IncrementPos(size);
```

Comment:
Check return value of Cast (and probably also IncrementPos)

Reply:
Good point!